### PR TITLE
[New] Measure distance in scene

### PR DESCRIPTION
**Pull request by CalebRas**
## Description

This PR implements `Measure distance in scene` in `Analysis` category.
URL to README: https://github.com/Esri/arcgis-maps-sdk-swift-samples/tree/Caleb/New-MeasureDistanceInScene/Shared/Samples/Measure%20distance%20in%20scene

## Linked Issue(s)

- `swift/issues/4158`

## How To Test

* Tap on the map to set the location distance measurement start location.
* Tap again to set the location distance measurement end location.
* Tap on "Imperial" to change the location distance measurement unit system.

## Screenshots

measure-distance-in-scene

## To Discuss

* As of PR creation, this sample does not implement any drag functionality.

### CODE REVIEW

**Comment by CalebRas** [comment]
@pgruenler I figured out the issue with points and added drag functionality so it should me good to go now. Let me know if you have any more problems.

**Comment by CalebRas** [comment]
@yo1995 Can you let me know if you think there is a better way to use the task blocks and if you think that this sample needs a view model?

**Comment by yo1995** [comment]
> Can you let me know if you think there is a better way to use the task blocks 

Will comment when I review.

> if you think that this sample needs a view model

This is a tough call. 4 objects (scene, analysis overlay, formatter, measurement) typically drive me towards a data model, but since the sample is only 150 lines I'll leave it up to you.

**Comment by CalebRas** [comment]
Sounds good. I ended up adding a view model.

File: Shared/Samples/Measure distance in scene/README.md
Comment (pgruenler) [inline Shared/Samples/Measure distance in scene/README.md:20]: I'm not exactly sure how to talk about these values but they are more of streamed properties in the Swift API rather that objects returned from a callback.

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (pgruenler) [inline Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift:151]:
```suggestion
            // Set scene to the viewpoint specified by the location distance measurement.
```

**Review by pgruenler — COMMENTED** [review]
Looks good and works well. Just a couple comments.

File: Samples.xcodeproj/project.pbxproj
Comment (yo1995) [inline Samples.xcodeproj/project.pbxproj:374]: Please remove the local toolkit reference from the project file.

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (yo1995) [inline Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift:176]: `common-samples/issues/3852`. If you like you can lend a help hand on figuring out why the data is down and help to replace affected designs. 🙂

File: Shared/Samples/Measure distance in scene/README.md
Comment (yo1995) [inline Shared/Samples/Measure distance in scene/README.md:26]: To add to Phil's comment, `MeasurementChangedEvent` is an internal symbol (LocationDistanceMeasurementMeasurementChangedEvent) so please remove it.

File: Shared/Samples/Measure distance in scene/measure-distance-in-scene.png
Comment (pgruenler) [inline Shared/Samples/Measure distance in scene/measure-distance-in-scene.png:1]: The latest screen shot has all zeros for the distances; I think it should show the correct distances. Is that a bug that the distances are wrong?

Comment (CalebRas) [reply Shared/Samples/Measure distance in scene/measure-distance-in-scene.png:1]: No, sorry. I must have tapped it.

**Review by pgruenler — DISMISSED** [review]

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (philium) [inline Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift:148]: Formatter objects [should not be used](https://developer.apple.com/documentation/foundation/data_formatting) in Swift:

> When working in Swift, use `formatted` methods directly on the types you want to format, optionally using [FormatStyle](https://developer.apple.com/documentation/foundation/formatstyle) and its subtypes to customize formatter output. This approach supports dates, integers, floating-point numbers, measurements, sequences, and person name components. Foundation caches identically-configured formatter instances internally, allowing you to focus on your app’s formatting needs.

Also, localized strings/values should not be stored. They should be computed on-demand when needed since (parts of) the locale can change while the app is running, affecting how the information is displayed. In this case, the measurements themselves should be stored and then formatted on demand like so:

```swift
Text("Direct: \(directDistance, format: ...)")
...
```

As for the format, one way you can reuse the same format for each measurement is by extending format style:

```swift
extension FormatStyle where Self == Measurement<UnitLength>.FormatStyle {
    static var distance: Self {
        .measurement(width: .abbreviated, usage: .asProvided, numberFormatStyle: .number.precision(.fractionLength(2)))
    }
}
```

And then passing `.distance` as the format: `format: .distance`.

Comment (CalebRas) [reply Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift]: @philium I updated the sample to use the format style, but can't get it work by storing the measurements. In using the following: 
```
for await measurements in model.locationDistanceMeasurement.measurements {
    directDistance = measurements.directDistance
    horizontalDistance = measurements.horizontalDistance
    verticalDistance = measurements.verticalDistance
}
```
The measurements updates come through the async stream just fine but in set them to the corresponding State variable, they wouldn't always update with the correct units when the unit system is toggled. Also, Changing the location distance measurement can cause some of distances to update when toggling the unit system but not all them. The aysnc stream and the API seem to be working fine but using `Measurement` State variables seems to be super buggy for some reason. Both Ting and looked at it and were at a loss on how to fix it so I just reverted back to using the localized strings. Let me know if there is better solution.

Comment (philium) [reply Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift]: That is surprising. I can't think of a better option at the moment, so what you have will have to do for now. Thank you for making those changes.

File: Samples.xcodeproj/project.pbxproj
Comment (yo1995) [inline Samples.xcodeproj/project.pbxproj:386]: Oops, please remove the local toolkit reference.

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (yo1995) [inline Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift:43]: > Can you let me know if you think there is a better way to use the task blocks

I tried a few ideas but since this sample supports both drag and single tap gestures, I can't find a good way to differentiate one from another without adding a few states. 

Thus I think in-situ task blocks are good.

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (yo1995) [inline Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift:61]:
```suggestion
                        if model.locationDistanceMeasurement.startLocation == model.locationDistanceMeasurement.endLocation {
                            return true
                        } else {
                            return false
                        }
```

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (yo1995) [inline Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift:142]:
```suggestion
            scene.initialViewpoint = Viewpoint(latitude: .nan, longitude: .nan, scale: .nan, camera: camera)
```
I don't like a point being the bounding geometry of a map or scene view as "bounding" sounds like an area or polygon to me. I like to explicitly set the unrelated parameter to `nan` to emphasize that a scene viewpoint can be solely determined by the camera, but it is up to you.

**Review by yo1995 — COMMENTED** [review]
Generally looking good!

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (philium) [reply Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift]: Or just `return model.locationDistanceMeasurement.startLocation == model.locationDistanceMeasurement.endLocation`.

File: Samples.xcodeproj/project.pbxproj
Comment (CalebRas) [reply Samples.xcodeproj/project.pbxproj]: Let me know if I didn't do it right.

**Review by pgruenler — APPROVED** [review]

**Review by yo1995 — APPROVED** [review]
Thanks for writing this sample! 

@Esri/swift-samples-reviewers After this PR is merged, Samples `v.next` depends on 200.2+ version.

File: Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift
Comment (yo1995) [reply Shared/Samples/Measure distance in scene/MeasureDistanceInSceneView.swift]: Phil's suggestion is good, let's apply that and add a comment to explain when the drag gesture is active.

**Review by yo1995 — DISMISSED** [review]

**Review by pgruenler — APPROVED** [review]

**Review by yo1995 — APPROVED** [review]

**Pull request merged**
